t/09-terminal.t

```diff
@@ -0,0 +1,340 @@
+#!/usr/bin/perl
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+use 5.018;
+use warnings;
+use Carp qw( confess );
+use English qw( -no_match_vars );
+use POSIX qw( :sys_wait_h pause );
+use Socket qw( PF_UNIX SOCK_STREAM sockaddr_un );
+use Time::HiRes qw( usleep );
+
+use Test::More tests => 12;
+
+BEGIN {
+    unshift @INC, '..';
+}
+
+use consoles::virtio_terminal;
+use testapi ();
+
+our $VERSION;
+
+my $log_path = './09-terminal-io.log';
+$testapi::password = 'd*97Jlk/.d';
+my $socket_path       = './virtio_console';
+my $login_prompt_data = <<'FIN.';
+
+
+Welcome to SUSE Linux Enterprise Server 12 SP2 RC3 (x86_64) - Kernel 4.4.21-65-default (hvc0).
+
+FIN.
+$login_prompt_data .= 'linux-5rw7 login: ';
+my $user_name_prompt_data = "login: ";
+my $user_name_data        = "root\n";
+my $password_prompt_data  = 'Password: ';
+my $password_data         = "$testapi::password\n";
+# Contains some ANSI/XTERM escape sequences
+my $first_prompt_data      = "\e[1mlinux-5rw7:~ #\e[0m\e(B";
+my $set_prompt_data        = qq/PS1="# "\n/;
+my $normalised_prompt_data = '# ';
+my $C0_EOT                 = "\cD\cD";
+my $C0_ETX                 = "\cC";
+my $C1_control_code        = qq(\eQ\n);
+my $US_keyboard_data       = <<'FIN.';
+!@\#$%^&*()-_+={}[]|:;"'<>,.?/~`
+abcdefghijklmnopqrstuwxyz
+ABCDEFGHIJKLMNOPQRSTUWXYZ
+0123456789
+FIN.
+my $stop_code_data        = "FIN.\n";
+my $repeat_sequence_count = 1000;
+my $next_test             = "GOTO NEXT\n";
+
+# If test keeps timing out, this can be increased or you can add more calls to
+# alarm in fake terminal
+my $timeout = 30;
+
+# Either write $msg to the socket or die
+sub try_write {
+    my ($fd, $msg) = @_;
+
+    open(my $logfd, '>>', $log_path);
+    print $logfd $msg;
+    close($logfd);
+
+  WRITE: while (1) {
```

Comment:
According to the Perl documentation I should be able to remove the while(1) and just use a plain block with a label, which I think is preferable because the intention is to only loop if EAGAIN happens, but it didn't work for some reason and left the loop when it shouldn't. I left the label there for readability, but I can remove it if it just makes people worried.